### PR TITLE
Feature/local node impl

### PR DESCRIPTION
**Pull request by sshaman1101**
This PR adds some local node methods.
Currently, we can implement Hub management operations and Marketplace operations.

TODO:

**Hub:**
- [x] Status

Workers:
- [x] WorkerList
- [x] WorkerStatus
- [x] GetWorkerProperties
- [x] SetWorkerProperties

ACLs:
- [x] Add stubs on Hub
- [x] GetRegistredWorkers
- [x] RegisterWorker
- [x] UnregisterWorker

Orders:
- [x] GetPlanByID
- [x] CreateAskPlan
- [x] DeleteAskPlan

Tasks:
- [x] TaskList
- [x] TaskStatus

**Market:**
- [x] GetOrders
- [x] GetOrderByID
- [x] CreateOrder (Bid only)
- [x] CancelOrder (Bid only)

### CODE REVIEW

**Comment by sshaman1101** [comment]
Implemented LocalNode methods for a ready to use parts of the SONM;
I see no reason to add not implemented commands to the CLI now (tasks and deals).

After this PR will be merged I would tune some inner logic into the Marketplace service.

@noxiouz @3Hren @antmat PTAL

File: Makefile [hunk 1]
```diff
@@ -82,6 +86,10 @@ install/cli: build/cli
 	@echo "+ $@"
 	cp ${CLI} ${INSTALLDIR}
 
+install/node: build/node
+	@echo "+ $@"
+	cp ${CLI} ${LOCAL_NODE}
```
Comment (antmat) [inline Makefile]: copy paste?

File: cmd/cli/commands/miner.go [hunk 2]
```diff
@@ -122,7 +115,7 @@ var minerStatusCmd = &cobra.Command{
 	PreRunE: checkHubAddressIsSet,
 	RunE: func(cmd *cobra.Command, args []string) error {
 		if len(args) < 1 {
-			return errMinerAddressRequired
+			return errWorkerAddressRequired
```
Comment (antmat) [inline cmd/cli/commands/miner.go]: if you already started renaming - how about something like 'errWorkerIdRequired' ?

File: hub.yaml [hunk 3]
```diff
@@ -1,13 +1,13 @@
 # An endpoint for miner connections.
 # allowed format is "ip:port" to bind Hub to given IP address
 # or ":port" to bind Hub to all available IPs
-endpoint: ":10002"
+endpoint: "172.16.1.118:10002"
```
Comment (antmat) [inline hub.yaml]: I don't think it's a good idea)

File: hub.yaml [hunk 4]
```diff
@@ -20,7 +20,7 @@ monitoring:
   # An endpoint for client connections using cli
   # allowed format is "ip:port" to bind Hub to given IP address
   # or ":port" to bind Hub to all available IPs
-  endpoint: ":10001"
+  endpoint: "172.16.1.118:10001"
```
Comment (antmat) [inline hub.yaml]: here too

File: insonmnia/node/common.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import "google.golang.org/grpc"
+
+// credentials describe any type of auth data.
+// Empty for now, but must be implemented in nearest future
+type credentials interface{}
+
+// initGrpcClient creates new client connection on addr and wraps it with credentials
+func initGrpcClient(addr string, _ credentials) (*grpc.ClientConn, error) {
+	cc, err := grpc.Dial(addr, grpc.WithInsecure(),
```
Comment (antmat) [inline insonmnia/node/common.go]: Another copy of this code. How about making a library method for connecting with sonm specific parameters?

File: insonmnia/node/hub.go [hunk 6]
```diff
@@ -1,52 +1,156 @@
...
 import (
+	log "github.com/noxiouz/zapctx/ctxlog"
 	pb "github.com/sonm-io/core/proto"
 	"golang.org/x/net/context"
 )
 
-type hubAPI struct{}
+type hubAPI struct {
+	conf Config
+	cc   pb.HubClient
```
Comment (antmat) [inline insonmnia/node/hub.go]: why 'cc'?

File: insonmnia/node/hub.go [hunk 7]
```diff
@@ -1,52 +1,156 @@
...
+	ctx  context.Context
+}
+
+func (h *hubAPI) Status(ctx context.Context, req *pb.Empty) (*pb.HubStatusReply, error) {
+	log.G(h.ctx).Debug("handling Status request")
```
Comment (antmat) [inline insonmnia/node/hub.go]: Should not it be 'Info'?

File: insonmnia/node/hub.go [hunk 8]
```diff
@@ -1,52 +1,156 @@
...
+
+	workers, err := h.getWorkersIDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	slots := []*pb.Slot{}
+	for _, wrkID := range workers {
+		req := &pb.ID{Id: wrkID}
+		workerSlots, err := h.cc.GetSlots(ctx, req)
```
Comment (antmat) [inline insonmnia/node/hub.go]: looks like this can be done on hub via one call(which would be more efficient)

File: insonmnia/node/hub.go [hunk 9]
```diff
@@ -1,52 +1,156 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	reply := &pb.TaskListReply{
+		Info: map[string]*pb.TaskListReply_TaskInfo{},
+	}
+
+	for _, wrkID := range workers {
+		taskStatuses, err := h.cc.MinerStatus(ctx, &pb.ID{Id: wrkID})
```
Comment (antmat) [inline insonmnia/node/hub.go]: here too - it looks like it would be better to add aggregating method to hub

File: node.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+# Marketplace service settings
+market:
+  # marketplace gRPC endpoint
+  endpoint: 127.0.0.1:9095
+
+# Hub management settings
+hub:
+  # endpoint is Hub's gRPC endpoint
+  # for client connections
+  endpoint: 172.16.1.118:10001
```
Comment (antmat) [inline node.yaml]: comment this and replace with some example like 1.2.3.4

File: cmd/cli/commands/common.go [hunk 11]
```diff
@@ -70,6 +83,13 @@ func checkHubAddressIsSet(cmd *cobra.Command, _ []string) error {
 	return nil
 }
 
+func checkNodeAddressIsSet(cmd *cobra.Command, _ []string) error {
```
Comment (antmat) [inline cmd/cli/commands/common.go]: Why do you need this when node address has default value?

File: cmd/cli/commands/interactor.go [hunk 12]
```diff
@@ -149,3 +149,224 @@ func NewGrpcInteractor(addr string, to time.Duration) (CliInteractor, error) {
...
+	_, err := it.market.CancelOrder(ctx, &pb.Order{Id: id})
+	return err
+}
+
+func (it *marketInteractor) ctx() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), it.timeout)
+}
+
+func NewMarketInteractor(addr string, timeout time.Duration) (NodeMarketInteractor, error) {
+	cc, err := grpc.Dial(addr, grpc.WithInsecure(),
```
Comment (antmat) [inline cmd/cli/commands/interactor.go]: again same code (and in many other places)

File: cmd/cli/commands/node_order.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+func loadOrderFile(path string) (*structs.Order, error) {
+	buf, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := task_config.OrderConfig{}
+	err = yaml.Unmarshal(buf, &cfg)
```
Comment (antmat) [inline cmd/cli/commands/node_order.go]: This code looks very similar to previous function (damn go with no generics!) 
Can we do something with it? something like loadYamlFileToStruct(path string, to interface{}) ?

File: cmd/cli/commands/node_tasks.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+				cmd.Printf("Worker \"%s\" has no tasks\r\n", worker)
+				continue
+			}
+
+			cmd.Printf("Worker \"%s\":\r\n", worker)
+			i := 1
+			for ID, status := range tasks.GetTasks() {
+				up := time.Duration(status.GetUptime())
+				cmd.Printf("  %d) %s \r\n     %s  %s (up: %v)\r\n",
+					i, ID, status.Status.String(), status.ImageName, up.String())
```
Comment (antmat) [inline cmd/cli/commands/node_tasks.go]: i++ in param?

File: cmd/cli/commands/node_worker.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+		_, err = hub.SetWorkerProperties(req)
+		if err != nil {
+			showError(cmd, "Cannot get workers status", err)
+			os.Exit(1)
+		}
+	},
+}
+
+func parsePropsFile(path string) (map[string]string, error) {
```
Comment (antmat) [inline cmd/cli/commands/node_worker.go]: again code duplication

File: cmd/cli/task_config/order.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+		ByuerID:    c.BuyerID,
+		SupplierID: c.SupplierID,
+		Price:      c.Price,
+		OrderType:  orderType,
+		Slot:       slot.Unwrap(),
+	})
+}
+
+func ParseOrderType(ty string) (sonm.OrderType, error) {
+	switch ty {
```
Comment (antmat) [inline cmd/cli/task_config/order.go]: I think this should be in structs package (and it can use generated protobuf map)

File: cmd/node/main.go [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+	n, err := node.New(ctx, cfg)
+	if err != nil {
+		fmt.Printf("Err: cannot build Node instance: %s\r\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Starting Local Node at %s...\r\n", cfg.ListenAddress())
+	if err := n.Serve(); err != nil {
+		fmt.Printf("Cannot start Local Node: %s\r\n", err.Error())
```
Comment (antmat) [inline cmd/node/main.go]: If some error happens during n.Serve() message will be a little misleading.

File: cmd/cli/commands/node_tasks.go (hunk 14 above)
Comment (sshaman1101) [reply cmd/cli/commands/node_tasks.go]: in param for what?

Comment (antmat) [reply cmd/cli/commands/node_tasks.go]: yep, you are right, go doesn't allow this

File: cmd/node/main.go (hunk 17 above)
Comment (sshaman1101) [reply cmd/node/main.go]: Now `.Serve()` method is just for serving :)

**Comment by 3Hren** [comment]
It's too hard to review such huge PR's. Either we review all and the entire process takes months, or merge it with hand-on-heart with further quick fix.

File: cmd/node/main.go (hunk 17 above)
Comment (antmat) [reply cmd/node/main.go]: I mean node is working for example for 1 day. After that it returns err and we see "Cannot start Local Node"

**Pull request merged**